packages/protocol/contracts/signal/SignalService.sol

```diff
@@ -71,113 +58,89 @@ contract SignalService is EssentialContract, ISignalService {
-            revert SS_INVALID_PARAMS();
-        }
-
-        Proof memory p = abi.decode(proof, (Proof));
-        if (!isMultiHopEnabled() && p.hops.length > 0) {
-            revert SS_MULTIHOP_DISABLED();
-        }
-
-        uint64 _srcChainId = srcChainId;
-        address _srcApp = app;
-        bytes32 _srcSignal = signal;
+        if (app == address(0) || signal == 0) revert SS_INVALID_PARAMS();
+
+        HopProof[] memory _hopProofs = abi.decode(proof, (HopProof[]));
+        if (_hopProofs.length == 0) revert SS_EMPTY_PROOF();
+
+        uint64 _chainId = chainId;
+        address _app = app;
+        bytes32 _signal = signal;
+        address _signalService = resolve(_chainId, "signal_service", false);
+
+        for (uint256 i; i < _hopProofs.length; ++i) {
+            HopProof memory hop = _hopProofs[i];
+
+            bytes32 signalRoot = _verifyHopProof(_chainId, _app, _signal, hop, _signalService);
+
+            bool isLastHop = i == _hopProofs.length - 1;
+            if (isLastHop) {
+                if (hop.chainId != block.chainid) revert SS_INVALID_LAST_HOP_CHAINID();
+                _signalService = address(this);
+            } else {
+                if (hop.chainId == 0 || hop.chainId == block.chainid) {
+                    revert SS_INVALID_MID_HOP_CHAINID();
+                }
+                _signalService = resolve(hop.chainId, "signal_service", false);
+            }
 
-        // Verify hop proofs
-        IHopRelayRegistry hrr;
-        if (p.hops.length > 0) {
-            hrr = IHopRelayRegistry(resolve("hop_relay_registry", false));
-        }
+            bool isFullProof = hop.accountProof.length > 0;
 
-        ICrossChainSync ccs = ICrossChainSync(resolve("taiko", false));
-        bytes32 stateRoot = ccs.getSyncedSnippet(p.height).stateRoot;
-
-        // If a signal is sent from chainA -> chainB -> chainC (this chain), we verify the proofs in
-        // the following order:
-        // 1. using chainC's latest parent's stateRoot to verify that chainB's TaikoL1/TaikoL2
-        // contract has sent a given hop stateRoot on chainB using its own signal service.
-        // 2. using the verified hop stateRoot to verify that the source app on chainA has sent a
-        // signal using its own signal service.
-        // We always verify the proofs in the reversed order (top to bottom).
-        for (uint256 i; i < p.hops.length; ++i) {
-            Hop memory hop = p.hops[i];
-            if (hop.stateRoot == stateRoot) revert SS_INVALID_HOP_PROOF();
-
-            if (!hrr.isRelayRegistered(_srcChainId, hop.chainId, hop.relay)) {
-                revert SS_INVALID_RELAY();
+            if (hop.cacheChainData) {
+                if (isLastHop) _relayChainData(_chainId, _SIGNAL_ROOT, signalRoot);
+                else if (isFullProof) _relayChainData(_chainId, _STATE_ROOT, hop.rootHash);
+                else _relayChainData(_chainId, _SIGNAL_ROOT, hop.rootHash);
             }
```

Comment:
```suggestion
            if (hop.cacheChainData && isFullProof) {
                _relayChainData(_chainId, _SIGNAL_ROOT, signalRoot);
            }
```

Mostly following up on my comment in the previous PR https://github.com/taikoxyz/taiko-mono/pull/15736#discussion_r1486860388 because I still believe we only have to cache signal roots.

Let's say you have to prove something on another L2 and the only thing that is available on the current chain is the L1 state root. To prove something like that we need the L2 account + storage proof (to prove the signal), and then the L1 account + storage proof (to prove the L2 state root that was used, which is made available by the taiko contract). 
- Current method: we would cache is the L2 state root and then the L1 signal root. If we next have to prove something on L1, we immediately have the L1 signal root so that's efficient. However, if we want to prove something on the L2 again, then the only thing we have is the L2 state root. So we again have to provide an account + storage proof to prove another signal, and now because of the special case on `isLastHop` we do cache the signal root so future signals on that L2 will now only have to provide a storage proof.
- Updated method: we would cache the L2 signal root and the L1 signal root. Now if we have to prove something on L2 again, we immediately have the necessary signal root available. So no need for another full proof to be able to cache the L2 signal root, and we also save on caching the the L2 state root which will never be used (the path to the L2 signal root is always shorter than the path to the L2 state root).

Generally speaking, I think the intermediate hops will always be full proofs, and for each of these we should always cache the signal root and the signal root only, so for future signals on those chains we can immediately depend on a trusted signal root. The last proof can be a partial proof or a full proof depending if one of the signal roots was cached previously. If the last proof is a partial proof, that means we verify against a signal root already and so need to cache anything anymore because it will already be written to the local signal service contract (otherwise a partial proof would not work).

Or perhaps I'm misunderstanding how it works, also a possibility.

Reply:
Main question still remains though: in what case is caching the state root more efficient so that somebody would choose to cache the state root over the signal root (or both)? There is a 1:1 mapping between the state root and the signal root, and for any signal that needs to be verified onchain we always need to prove the signal root for a certain state root on any chain (and any hops used). So caching the state root doesn't seem useful to me for proving signals, because for any signal that is being proven, there already needs to be an account proof onchain that verifies the signal root as well to do anything with it. And if only the state root is cached, that same account proof would have to be put onchain again and verified in a following signal proof, so would be less efficient in all cases.

Of course, if the argument is that the state root being cached could in theory be useful for other dapps that need to read data from other contracts that are not the signal contract, it could indeed be useful! So if that is the expected use case of being able to cache state roots than fair enough! But then there's still the problem that the caching is only done through proving signals, so only people that are interested in making the proving of signals more efficient will actually cache any data, so at least for now there doesn't seem to be any incentive to cache any state roots even in that case as far as I can see.